File: benchmarks/python/records.py
from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass


@dataclass
class Point:
    x: int
    y: int


@dataclass
class Rectangle:
    corner: Point
    width: int
    height: int


def area(r: Rectangle) -> int:
    return r.width * r.height


def move_right(p: Point, amount: int) -> None:
    p.x = p.x + amount


def main() -> None:
    origin = Point(x=0, y=0)
    frame = Rectangle(corner=deepcopy(origin), width=4, height=3)
    print(f"area: {area(frame)}")

    move_right(origin, 10)
    print(f"origin moved to x {origin.x}")
    print(f"frame's corner is still at x {frame.corner.x}")


if __name__ == "__main__":
    main()

File: benchmarks/rust/records.rs
#[derive(Clone)]
#[allow(dead_code)]
struct Point {
    x: i64,
    y: i64,
}

struct Rectangle {
    corner: Point,
    width: i64,
    height: i64,
}

fn area(r: &Rectangle) -> i64 {
    r.width * r.height
}

fn move_right(p: &mut Point, amount: i64) {
    p.x = p.x + amount;
}

fn main() {
    let mut origin = Point { x: 0, y: 0 };
    let frame = Rectangle {
        corner: origin.clone(),
        width: 4,
        height: 3,
    };
    println!("area: {}", area(&frame));

    move_right(&mut origin, 10);
    println!("origin moved to x {}", origin.x);
    println!("frame's corner is still at x {}", frame.corner.x);
}
